### DISCUSSION TITLE
Why does pyright verifytypes state that TypeAlias imported is unknown?

### DISCUSSION THREAD

**Discussion by GoogolGenius** [microsoft/pyright#3603] [Q&A] [answered]
I have a TypeAlias in another file that looks like this:
```python
TaskCallbackT = Callable[[], Awaitable[None]]
```

However running `pyright --verifytypes` creates errors, one of which is displayed below:
```
Symbols used in public interface:
tasks.cron.Cronjob
  error: Type of metaclass unknown
  error: Type of base class unknown
tasks.cron.Cronjob.__init__
  c:\Users\_\Documents\_\GitHub\hikari-crescent\crescent\ext\tasks\cron.py:13:9 - error: Type of parameter "callback" is unknown
crescent.ext.tasks.cron.cronjob
  c:\Users\_\Documents\_\GitHub\hikari-crescent\crescent\ext\tasks\cron.py:24:5 - error: Return type is partially unknown
    Return type is "(Unknown) -> MetaStruct[Unknown, Cronjob]"
      Return type is partially unknown
        Return type is "MetaStruct[Unknown, Cronjob]"
          Type argument 1 for class "MetaStruct" has unknown type
```

This happens when I import the file with the TypeAlias into another to use the type. If I directly put the actual type hint, it works as expected. Is this a bug?

**Comment by erictraut** [comment, marked as answer]
Are you sure you want to be using the `--verifytypes` option? That's meant for library authors who are writing "py.typed" libraries and want to validate that the public interface to the library is fully typed. If your intent is to simply type check your code, you should not use this option.

If your intent is to use `--verifytypes`, then I'll need more context to answer your question. I can't tell from the error how you are using the type alias.

**Reply by GoogolGenius** [reply]
I am currently working on improving pyright support for a py.typed library. So yes, I probably want to use the `--verifytypes` option. 
This is the subpackage directory where it's used: https://github.com/GoogleGenius/hikari-crescent/tree/task/improve-pyright-support/crescent/ext/tasks

`TaskCallbackT = Callable[[], Awaitable[None]]` is defined in the task.py which is used there for type hints, then exported from the tasks package defined in the \_\_init\_\_.py and also imported for use in the other modules of the tasks subpackage as type hints. 

I also still got these errors if I changed every TypeAlias typehint to the actual typehint: 
```
Symbols used in public interface:
tasks.cron.Cronjob
  error: Type of metaclass unknown
  error: Type of base class unknown
tasks.loop.Loop
  error: Type of metaclass unknown
  error: Type of base class unknown
```
The classes go like so:
```python
class Cronjob(Task):
    def __init__(self, cron: str, callback: TaskCallbackT) -> None:
        self.cron: croniter = croniter(cron, datetime.now())

        super().__init__(callback)

    def _next_iteration(self) -> float:
        call_next_at: datetime = self.cron.get_next(datetime)
        time_to_next = call_next_at - datetime.now()
        return time_to_next.total_seconds()
```
```python
class Loop(Task):
    def __init__(self, callback: TaskCallbackT, delay_seconds: float) -> None:
        self.delay_seconds = delay_seconds
        self.first_loop: bool = True

        super().__init__(callback)

    def _next_iteration(self) -> float:
        """
        Returns 0 if the first loop has not occurred, otherwise return `Loop.delay_seconds`.
        """
        if self.first_loop:
            return 0
        return self.delay_seconds

    def _call_next(self) -> None:
        super()._call_next()
        self.first_loop = False
```
```python
class Task(ABC):
    def __init__(self, callback: TaskCallbackT) -> None:
        self.event_loop: AbstractEventLoop = get_event_loop()
        self.callback = callback
        self.timer_handle: TimerHandle | None = None
        self.app: Bot | None = None

    def start(self) -> None:
        if self.running:
            raise TaskError("Task is already running.")
        ensure_future(self._start_inner())

    async def _start_inner(self) -> None:
        assert self.app is not None

        await self.app.started.wait()

        self._call_next()

    def stop(self) -> None:
        if self.timer_handle:
            self.timer_handle.cancel()
    
    # more is below in the actual file (crescent.ext.tasks.task.Task), just omitted for space
```

**Reply by erictraut** [reply]
Are you running `--verifytypes` on the installed library? In other words, have you run setup on the library, and it is now in your `site_packages` directory? The `--verifytypes` option is designed to examine the library in its installed location. This is important because some library modules (like yours) use absolute imports, so correct import resolution requires them to be installed in their intended location.

The errors you're reporting above don't appear to have anything to do with the `TaskCallbackT` type alias. The error `error: Type of base class unknown` means that one or more base classes for the `Cronjob` class has an unknown type. Since `Cronjob` derives only from `Task`, that means pyright isn't able to resolve the type of `Task` in that case.

It's difficult for me to diagnose this further without a complete set of repro steps.

**Reply by GoogolGenius** [reply]
Repro:
- Clone https://github.com/GoogleGenius/hikari-crescent.git
- Switch to `task/improve-pyright-support` branch
- `pip install .`
- Pyright v1.1.255
- `pyright --verifytypes crescent` (also tried replacing "crescent" to the full path of the lib in the venv as well)

Decided to put the full pyright run in a pastebin instead: https://mystb.in/NorfolkCommunicateSet.css

**Reply by erictraut** [reply]
Thanks for the repro steps. I was able to repro the issue.

The problem is that you are missing a `__init__.py` file within the `ext` subdirectory. Because it is missing this file, it is considered a implicit namespace rather than a traditional module, and PEP 561 indicates that namespaces must have their own "py.typed" files — i.e. they don't inherit a "py.typed" from an outer module. Various packages rely on this behavior, so it's not something we can change.

So the correct fix here is to add an empty `__init__.py` file within the `ext` directory so `ext` is seen as a traditional submodule. It will then inherit the "py.typed" from its parent module.

It's not surprising that you weren't able to come to this conclusion on your own based on the error messages pyright was providing. There must be a better way to report this error, but I haven't come up with a good solution yet. I'll keep thinking about it.

**Reply by GoogolGenius** [reply]
Ah, that makes sense, yeah it wasn't clear to me that `__init__.py` would be needed there. Thank you so much for the help!
